drgn_tools/dm.py

```diff
@@ -0,0 +1,118 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
+# Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/
+"""
+Helpers for device mapper devices.
+"""
+import argparse
+from typing import Iterable
+from typing import Tuple
+
+from drgn import Object
+from drgn import Program
+from drgn.helpers.linux.list import list_for_each_entry
+from drgn.helpers.linux.rbtree import rbtree_inorder_for_each_entry
+
+from drgn_tools.corelens import CorelensModule
+from drgn_tools.module import ensure_debuginfo
+from drgn_tools.table import print_table
+from drgn_tools.util import BitNumberFlags
+from drgn_tools.util import kernel_version
+from drgn_tools.util import variable_exists
+
+
+def for_each_dm_hash(prog: Program) -> Iterable[Tuple[Object, str]]:
+    # bucket size is 64
+    for i in range(64):
+        head = prog["_name_buckets"][i]
+        for hc in list_for_each_entry(
+            "struct hash_cell", head.address_of_(), "name_list"
+        ):
+            yield hc.md, hc.name.string_().decode()
+
+
+def for_each_dm_rbtree(prog: Program) -> Iterable[Tuple[Object, str]]:
+    for hc in rbtree_inorder_for_each_entry(
+        "struct hash_cell", prog["name_rb_tree"], "name_node"
+    ):
+        yield hc.md, hc.name.string_().decode()
+
+
+def for_each_dm(prog: Program) -> Iterable[Tuple[Object, str]]:
+    if variable_exists(prog, "_name_buckets"):
+        dm_list = list(for_each_dm_hash(prog))
+    elif variable_exists(prog, "name_rb_tree"):
+        dm_list = list(for_each_dm_rbtree(prog))
+
+    for dm, name in dm_list:
+        yield dm, name
+
+
+class dm_flags_bits(BitNumberFlags):
+    """
+    Class to convert preprocessor definitions to enum
+
+    drgn can't get the value of preprocessor definitions.
+    This is only appliable to the kernel starting 8ae126660fdd
+    which was merged by v4.10
+    """
+
+    BLOCK_IO_FOR_SUSPEND = 0
+    SUSPENDED = 1
+    FROZEN = 2
+    FREEING = 3
+    DELETING = 4
+    NOFLUSH_SUSPENDING = 5
+    DEFERRED_REMOVE = 6
+    SUSPENDED_INTERNALLY = 7
+    POST_SUSPENDING = 8
+    EMULATE_ZONE_APPEND = 9
+
+
+class dm_flags_bits_old(BitNumberFlags):
+    """only appliable to kernel older than v4.10"""
+
+    BLOCK_IO_FOR_SUSPEND = 0
+    SUSPENDED = 1
+    FROZEN = 2
+    FREEING = 3
+    DELETING = 4
+    NOFLUSH_SUSPENDING = 5
+    DMF_MERGE_IS_OPTIONAL = 6
+    DMF_DEFERRED_REMOVE = 7
+    DMF_SUSPENDED_INTERNALLY = 8
```

Comment:
I'm guessing you left the prefix `DMF_` on these in order to distinguish these flag names from the new flags. If you want that distinction visible in the table, then that's fine.

But it's easy to refer to both versions of the flags when they have the same name, e.g. `DmFlagBitsOld.DEFERRED_REMOVE` vs `DmFlagBits.DEFERRED_REMOVE`. So if you don't care about the string output, I'd prefer for all of these constants to appear without the `DMF_` prefix.